### PR TITLE
Allow switching language using lang url param

### PR DESCRIPTION
**Pull request by addshore**
Currently if a language is selected and i18n messages
are missing the i18n code will be displayed instead

### CODE REVIEW

**Comment by jakobw** [comment]
This approach seems a bit hacky to me. Can't we take the remaining strings from the `index.html`, put them in the `i18n.json` and then do the language switching entirely in JS?

**Comment by manicki** [comment]
I was actually also thinking about something like what @jakobw suggested. I am not sure which solution will be nicer though.

**Comment by addshore** [comment]
Well assuming not all browsers have JS we would probably always went a landing page which also renders useful content without JS.
So if we want the landing page to also be localizable this would have to be done in PHP.

Fixing the landing page for non JS browsers is also something that should be done in my opinion saying that the site requires JS, right now it just half loads and doesn't work.

**Comment by manicki** [comment]
That's a good point. There should be some kind of i18n-enabled noscript info.

**Comment by addshore** [comment]
I went ahead and moved the i18n files out of the JS directory (as they are used by the bit of php now)

I also went ahead and created a basic noscript landing page.
It looks as below:

image

(although the message needs translating to German and adding to the i18n file)

**Comment by addshore** [comment]
My current thoughts for the large blocks of text on the landing page are that they could be split out into separate html files which could then each be localized?

./i18n/html/de/about.html
./i18n/html/de/legalt.html

The text on feedback is probably small enough to keep in the json file but the about and legal sections just seems a bit insane.

**Comment by addshore** [comment]
As far as I can tell this PR now covers i18n for all strings (except internal exception messages which are not really user facing).

**Comment by jakobw** [comment]
Ohh, that looks really cool now!

**Comment by jakobw** [comment]
If I remember correctly I also suggested putting large blocks of texts in files so they can be translated but somehow after we discussed this we concluded that it would be better to have them all in one file (split large texts into paragraphs) to make it easier to just give it to people for translation. I think this looks pretty good though and it's super nice that we can even have a translated message for people who don't have JS activated. What do @manicki and @tobijat think? :)

File: backend/src/Actions/FeedbackAction.php [hunk 1]
```diff
@@ -5,22 +5,31 @@
...
+		$request = $app['request'];
+
+		$lang = !empty( $request->get( 'lang' ) ) ? $request->get( 'lang' ) : 'de';
+		$langFile = __DIR__ . '/../../../i18n/' . $lang . '/i18n.json';
+		if( !file_exists( $langFile ) ) {
+			$langFile = __DIR__ . '/../../../i18n/de/i18n.json';
+		}
+		$this->i18n = json_decode( file_get_contents( $langFile ), true );
 
 		if ( $this->requestValid( $app['request'] ) ) {
```
Comment (jakobw) [inline backend/src/Actions/FeedbackAction.php]: Could also be `$this->requestValid( $request )`

File: index.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+// If we have any i18n data replace the i18n codes with new strings
+if ( $i18nData && isset( $i18nData['index'] ) ) {
+    foreach( $i18nData['index'] as $key => $string ) {
+        $html = str_replace( 'i18n.index.' . $key, $string, $html );
+    }
+}
+
+// Also replace html snippets
+$htmlFiles = array(
```
Comment (jakobw) [inline index.php]: We can haz new array `[]` syntax!

File: backend/src/Actions/FeedbackAction.php (hunk 1 above)
Comment (addshore) [inline backend/src/Actions/FeedbackAction.php]: Done

File: index.php (hunk 2 above)
Comment (addshore) [inline index.php]: Will do everywhere! :)

**Comment by addshore** [comment]
Yes the longer texts in files do seem quite manageable, however I feel that I should take a look at legal.html and remove the tracking switch from there....

I guess the switch itself should remain in the main html file and the strings around there should head to the i18n.json file

File: js/app/Messages.js [hunk 3]
```diff
@@ -1,6 +1,32 @@
 'use strict';
 
 var Polyglot = require( 'node-polyglot' ),
-	i18n = require( '../i18n.json' );
+	i18n;
 
-module.exports = new Polyglot( { phrases: i18n[ 'de' ] } );
+var getParam = function getUrlParameter( sParam ) {
```
Comment (jakobw) [inline js/app/Messages.js]: I thought JS would have a built-in thing to deal with URLs but google didn't spit anything out :/

File: i18n/de/i18n.json [hunk 4]
```diff
@@ -0,0 +1,106 @@
+{
+  "index": {
+    "title": "Lizenzhinweisgenerator",
+    "description": "Lizenzhinweise für Bilder aus Wikipedia und Wikimedia Commons",
+    "no-script": "Sie müssen JavaScript zur Nutzung dieser Website aktiviert.",
```
Comment (manicki) [inline i18n/de/i18n.json]: I am guessing this is not entirely correct Kraut speak, although it contains all words I'd expect: "müssen", "Javascript", "aktiviert" :)
Let's have @jakobw and @WMDE-Fisch suggest something better

Comment (jakobw) [inline i18n/de/i18n.json]: Haha, close enough though. Something like "Zur Nutzung dieser Webseite muss JavaScript aktiviert sein." would be better.

File: index.base.html [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+      <noscript>
+        <div class="no-script" >
+          <h2>{{i18n.index.no-script}}</h2>
+        </div>
+      </noscript>
+
+      <div class="attribution" id="attribution">
+        <noscript>
+          ""<a href="https://commons.wikimedia.org/wiki/User:The_Photographer">The Photographer</a>, <a href="https://commons.wikimedia.org/wiki/File:A_day_of_fishing_aground.jpg">„A day of fishing aground“</a>, <a href="https://creativecommons.org/publicdomain/zero/1.0/legalcode">CC0 1.0</a>"
```
Comment (manicki) [inline index.base.html]: there should probably be a single opening quote at the beginning of the line here

**Comment by manicki** [comment]
Looks really nice!
I really like having translatable noscript information and splitting i18n json files into separate file per language.
There is a little bit of inconsistency ie. when some `i18n.index` translations are missing there is no German fallback. For instance if I create an empty `i18n/en/i18n.json` file all HTML stuff will be in German same as all JS generated text but i18n.index things will remain untranslated, ie. it will be "{{i18n.index.title}}" instead of "Lizenz..." displayed on the front screen.
But I don't think it is a problem as it will only occur where there are some particular translations missing which can actually be helpful in the process of adding new language etc as it is obvious what is still not translated.

So basically it will be ready to go after updating one translation to what Jakob suggested.

**Comment by WMDE-Fisch** [comment]
+:100:
